### PR TITLE
Data preprocess/bill

### PR DESCRIPTION
**Pull request by billyang98**
set up the entire dataset with processing script and added how to use in the training loop

### CODE REVIEW

File: src/prepare_data.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        sample = {'mel':mel, 'mag': mag}
+
+        return sample
+
+if __name__ == '__main__':
+    dataset = PrepareDataset(os.path.join(hp.data_path,'metadata.csv'), os.path.join(hp.data_path,'wavs'))
+    dataloader = DataLoader(dataset, batch_size=1, drop_last=False, num_workers=8)
+    from tqdm import tqdm
+    pbar = tqdm(dataloader)
+    for d in pbar:
```
Comment (Lucaskabela) [inline src/prepare_data.py:45]: Is this just so __getitem__ gets called on each wav? If so, is it really necessary to wrap this in a Pytorch DataLoader/Dataset object?

File: src/preprocess.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+    def __getitem__(self, idx):
+        wav_name = os.path.join(self.root_dir, self.landmarks_frame.ix[idx, 0]) + '.wav'
+        original_text = self.landmarks_frame.ix[idx, 1]
+
+        text = np.asarray(raw_text_to_phoneme_ids(original_text), dtype=np.int32)
+        mel = np.load(wav_name[:-4] + '.pt.npy')
+        mel_input = np.concatenate([np.zeros([1,hp.num_mels], np.float32), mel[:-1,:]], axis=0)
+        text_length = len(text)
+        pos_text = np.arange(1, text_length + 1)
```
Comment (Lucaskabela) [inline src/preprocess.py:41]: What exactly are the "pos_*" variables used for?  Is this for our network output?

File: src/utils.py [hunk 3]
```diff
@@ -1,3 +1,51 @@
 '''
 Contains any and all code we didnt want to put somewhere else
-'''
\ No newline at end of file
+'''
+import numpy as np
+import librosa
+import hyperparameters as hp
```
Comment (Lucaskabela) [inline src/utils.py]: What are our thoughts on hyperparameters?  I was considering using a txt file and arg parse, but I kind of like the simplicity of just having a python file with them...

File: src/prepare_data.py (hunk 1 above)
Comment (billyang98) [reply src/prepare_data.py:45]: Yes, it is not. This is just copied over from the old repo. The code is terrible but there's not too much value in refactoring this code. It is only called once to make the mel files before training on a single machine.

File: src/preprocess.py (hunk 2 above)
Comment (billyang98) [reply src/preprocess.py:41]: They are indices basically. They are left over from the other repo. Could be useful for training? But I think they can mostly be ignored. Thought it'd be nice to leave in case we use it in the future.

File: src/utils.py (hunk 3 above)
Comment (billyang98) [reply src/utils.py]: I added support for a json file for args. The python file is only good for consts, so I renamed it to audio_parameters, and moved actual training hyperparameters to train.json as an example. That way we can just make new json files when training under new parameters.

Comment (Lucaskabela) [reply src/utils.py]: Love it!

**Review by Lucaskabela — APPROVED** [review]

File: src/prepare_data.py (hunk 1 above)
Comment (liaojh1998) [reply src/prepare_data.py:45]: I think the only other reason they use DataLoader is so that there are multiple num_worker sub-processes to speed it up.

This looks good, I used this before the dataset as well.

File: src/preprocess.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    def __len__(self):
+        return len(self.landmarks_frame)
+
+    def __getitem__(self, idx):
+        wav_name = os.path.join(self.root_dir, self.landmarks_frame.ix[idx, 0]) + '.wav'
+        mel = np.load(wav_name[:-4] + '.pt.npy')
+        mag = np.load(wav_name[:-4] + '.mag.npy')
+        sample = {'mel':mel, 'mag':mag}
+
+        return sample
```
Comment (liaojh1998) [inline src/preprocess.py:70]: What's the purpose of the PostDataset? Is this another dataset or is this to be used with the test set?

File: src/preprocess.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    return np.pad(inputs, [[0,0],[0,0],[0, ap.outputs_per_step - (timesteps % ap.outputs_per_step)]], mode='constant', constant_values=0.0)
+
+def get_param_size(model):
+    params = 0
+    for p in model.parameters():
+        tmp = 1
+        for x in p.size():
+            tmp *= x
+        params += tmp
+    return params
```
Comment (liaojh1998) [inline src/preprocess.py:139]: This is not used anywhere.

Should we move this to `utils.py`, or should we keep it here so that when we copy code over, it's easy to keep the imports?

**Review by liaojh1998 — APPROVED** [review]
Thank you for working on this. There are a few is-this-needed questions from me but I think this is good to go.

File: src/preprocess.py (hunk 4 above)
Comment (billyang98) [reply src/preprocess.py:70]: not needed, was used in the old repo, i just copied and didn't delete it

File: src/preprocess.py (hunk 5 above)
Comment (billyang98) [reply src/preprocess.py:139]: I fully copied things over so it would be easier to keep a record of the actual changes made. This probably can get removed, but let's leave it for now in case.
